sale_delivery_block/models/sale_order.py

```diff
@@ -0,0 +1,71 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Eficent Business and IT Consulting Services S.L.
+#   (http://www.eficent.com)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, fields, models, _
+from odoo.exceptions import ValidationError
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    @api.multi
+    @api.constrains('delivery_block_id')
+    def _check_not_auto_done(self):
+        for so in self:
+            if so.delivery_block_id and self.env['ir.values'].get_default(
+                    'sale.config.settings', 'auto_done_setting'):
+                raise ValidationError(
+                    _('You cannot block a sale order with "auto_done_setting" '
+                      'active.'))
+
+    delivery_block_id = fields.Many2one(
+        comodel_name='sale.delivery.block.reason', track_visibility='always',
+        string='Delivery Block Reason', readonly=True,
+        states={'draft': [('readonly', False)], 'sent': [('readonly', False)]})
+
+    @api.multi
+    @api.onchange('partner_id')
+    def onchange_partner_id(self):
+        """Add the 'Default Delivery Block Reason' if set in the partner."""
+        res = super(SaleOrder, self).onchange_partner_id()
+        for so in self:
+            so.delivery_block_id = so.partner_id.default_delivery_block or \
+                False
+        return res
+
+    @api.multi
+    def action_remove_delivery_block(self):
+        """Remove the delivery block and create procurements as usual."""
+        self.write({'delivery_block_id': False})
```

Comment:
Maybe you can add before a filter (the same as in the view) as you cannot rely logic just on form view side (remember that this can be called through xml-rpc).